### ISSUE TITLE
Ability to change vSphere VM hardware config as part of IPI deployment

### ISSUE THREAD

**Issue by saintdle** [openshift/installer#4613]
# Version

4.6 and later

# Platform:

vSphere

* IPI 

# What happened?

When deploying using OpenShift-Install command, it would be great to be able to specify the hardware versions of the virtual machines deployed in vSphere. 

This is necessary to use things like the [VMware CSI ](https://github.com/mylesagray/ocp-vsphere-csi)which required hardware v15, the RHCOS OVA is hardware v13. 

Which means once a cluster is deployed, everything needs to be powered off to update the hardware version in vSphere.

# What you expected to happen?

The ability to set the hardware version as a line in the install-config.yaml file before you run "openshift-install create cluster"

This can then be achieved by uploading the RHCOS template VM, but modifying the hardware level before it is cloned it the control-plane and compute-plane virtual machines.

**Comment by staebler** [comment]
@jcpowermac Can you shed some light on this? The OpenShift docs for [installing on vSphere](https://docs.openshift.com/container-platform/4.6/installing/installing_vsphere/installing-vsphere-installer-provisioned.html#installation-vsphere-infrastructure_installing-vsphere-installer-provisioned) indicate that we support storage with the CSI driver when using vSphere 6.7U3 or higher, but the docs do not mention anything about changing in the hardware version that I can find.

**Comment by saintdle** [comment]
> @jcpowermac Can you shed some light on this? The OpenShift docs for [installing on vSphere](https://docs.openshift.com/container-platform/4.6/installing/installing_vsphere/installing-vsphere-installer-provisioned.html#installation-vsphere-infrastructure_installing-vsphere-installer-provisioned) indicate that we support storage with the CSI driver when using vSphere 6.7U3 or higher, but the docs do not mention anything about changing in the hardware version that I can find.

https://vsphere-csi-driver.sigs.k8s.io/driver-deployment/prerequisites.html#virtual-machine-configuration-

> VM Hardware version must be 15 or higher.

Also logged here on this Red Hat blog post

https://www.openshift.com/blog/installing-and-configuring-vsphere-csi-driver-on-openshift-4.3

**Comment by staebler** [comment]
Yes, I see that the hardware version must be 15 or higher. I have no contention with that point. I am looking for Joseph's input with regards to how best to handle this, because his vSphere knowledge and experience is wider and deeper than mine.
1. Does it make any sense to produce a different OVA for the different target hardware versions?
2. Should the installer query the vCenter in order to deduce the appropriate hardware version to use?
3. What is the user impact here with respect to number of users impacted and burden imposed on the user by having it as a day-2 operation as opposed to an installer-supported configuration?

**Comment by jcpowermac** [comment]
Since we still support 6.5 - 6.7U2 anything we change needs to take that into account.

> @jcpowermac Can you shed some light on this? The OpenShift docs for [installing on vSphere](https://docs.openshift.com/container-platform/4.6/installing/installing_vsphere/installing-vsphere-installer-provisioned.html#installation-vsphere-infrastructure_installing-vsphere-installer-provisioned) indicate that we support storage with the CSI driver when using vSphere 6.7U3 or higher, but the docs do not mention anything about changing in the hardware version that I can find.

Currently the in-tree storage provider is still installed. We support the out-of-tree CSI just like any alternative driver but if that CSI was the cause of a problem then VMware would need to be contacted.

> Does it make any sense to produce a different OVA for the different target hardware versions?

We could modify the hardware on import of the OVA or when creating the virtual machines. It probably makes more sense when the template is imported since MAO would be cloning it for workers as well. Its fairly trivial to upgrade the virtual hardware vs. the potential complexity of multiple ova(s).

> Should the installer query the vCenter in order to deduce the appropriate hardware version to use?

That would be really cool. It would need to determine the ESXi host version for the nodes in the vCenter cluster and then take the lowest version.

See: https://issues.redhat.com/browse/CORS-1570

> What is the user impact here with respect to number of users impacted and burden imposed on the user by having it as a day-2 operation as opposed to an installer-supported configuration?

1. power off vm
2. Upgrade via CLI(s):
```
Get-VM <vm> | Set-VM -Version v15 -Confirm:$False
```
```
govc vm.upgrade 
```
3. repeat

I think day two currently is fine. Once the storage team includes the out-of-tree CSI the installer should be upgrading the virtual hardware to the appropriate version.

**Comment by saintdle** [comment]
> Yes, I see that the hardware version must be 15 or higher. I have no contention with that point. I am looking for Joseph's input with regards to how best to handle this, because his vSphere knowledge and experience is wider and deeper than mine.
> 
>     1. Does it make any sense to produce a different OVA for the different target hardware versions?

No, I think in your OpenShift-Installer you should build something where the user can provide a flag to set the Hardware version they want in their environment. This becomes more flexible in the future, as VMware will release newer vSphere versions with new HW versions which add additional capabilities (such as direct IO path to hardware for example).

>     2. Should the installer query the vCenter in order to deduce the appropriate hardware version to use?
You could do this but the logic would become a little messy.

>     3. What is the user impact here with respect to number of users impacted and burden imposed on the user by having it as a day-2 operation as opposed to an installer-supported configuration?

The burden is having to power off your cluster and manage this. When it's a simple configuration which could be taken care of within the OpenShift-Install tool. 

Also as you noted, your Red Hat documentation needs to add the note in about needing Hardware version 15.

**Comment by jcpowermac** [comment]
> Also as you noted, your Red Hat documentation needs to add the note in about needing Hardware version 15.

We are using the in-tree CSI so there is no reason currently to update documentation. If a customer would like to install the out-of-tree driver they need to follow the steps in the upstream documentation. Once the out-of-tree is included directly in OpenShift yes I agree the documentation should be updated.

**Comment by RobbieJVMW** [comment]
I believe a much more elegant approach would be to default to the latest hardware level the cloud provider (vSphere) permits.  This allows the hypervisor platform to take full advantage of the VM's provisioning against it and allows OCP is deliver optimum performance.  The OCP platform then would be able to take advantage of any out of tree drivers without having VI Admin involvement to power down and version change the virtual hardware level.    

It should also be noted that as part of regular vSphere platform maintenance it maybe policy to have all the VM's running at the latest hardware level which can be forced by the VI Admin.  This would cause a reboot of the OCP VM's and potential service outage.

**Comment by dav1x** [comment]
I just tested. If you remove the following line from the OVF file:

```
      <System>
        <vssd:ElementName>Virtual Hardware Family</vssd:ElementName>
        <vssd:InstanceID>0</vssd:InstanceID>
        <vssd:VirtualSystemIdentifier>image</vssd:VirtualSystemIdentifier>
        **<vssd:VirtualSystemType>vmx-07 vmx-08</vssd:VirtualSystemType>**
      </System>
```

This import OVF template defaults to whatever level the vSphere host supports. I deployed a 6.5U2 cluster and the VM imported as:

![Screen Shot 2021-02-22 at 11 43 00 AM](https://user-images.githubusercontent.com/7294149/108747562-2e798680-7503-11eb-8e4e-2631de18e145.png)

The same OVF template imported into a vSphere 7.0 host imports as:

![Screen Shot 2021-02-22 at 11 44 00 AM](https://user-images.githubusercontent.com/7294149/108747633-45b87400-7503-11eb-82f6-f0c3f1e62838.png)

This may be an easy fix.
